dewy/common/collection_embeddings.py

```diff
@@ -0,0 +1,259 @@
+from typing import List, Self, Tuple
+
+import asyncpg
+from llama_index.embeddings import OpenAIEmbedding
+from llama_index.node_parser import SentenceSplitter
+from llama_index.schema import TextNode
+from loguru import logger
+
+from dewy.chunks.models import TextChunk
+from dewy.collections.models import DistanceMetric
+from dewy.collections.router import get_dimensions
+
+from .extract import extract
+
+
+class CollectionEmbeddings:
+    """Helper class for working with the embeddings in a collection."""
+
+    def __init__(self, pg_pool: asyncpg.Pool, collection_row: asyncpg.Record) -> None:
+        """Create a new CollectionEmbeddings.
+
+        Parameters:
+        - pg_pool: The asyncpg pool for connecting to the database
+        - collection_row: A record from the database containing the following
+          fields: collection_id, name, text_embedding_model, text_distance_model.
+        """
+        self._pg_pool = pg_pool
+
+        self.collection_id = collection_row["collection_id"]
+        self.name = collection_row["name"]
+        self.text_embedding_model = collection_row["text_embedding_model"]
+        self.text_distance_metric = DistanceMetric(
+            collection_row["text_distance_metric"]
+        )
+        self.extract_tables = False
+        self.extract_images = False
+
+        # TODO: Look at a sentence window splitter?
+        self._splitter = SentenceSplitter()
+        # TODO: Support other embeddings (based on the model).
+        self._embedding = OpenAIEmbedding()
+
+        # TODO: Figure out how to limit by the number of *chunks* not the number
+        # of embeddings.
+        dimensions = get_dimensions(self.text_embedding_model)
+        field = f"embedding::vector({dimensions})"
+
+        self._retrieve_embeddings = f"""
+        SELECT
+          chunk_id,
+          {self.text_distance_metric.distance(field, "$1")} AS score
+        FROM embedding
+        WHERE collection_id = {self.collection_id}
+        ORDER BY {self.text_distance_metric.order_by(field, "$1")}
+        LIMIT $2
+        """
+
+        self._retrieve_chunks = f"""
+        WITH relevant_embeddings AS (
+          SELECT
+            chunk_id,
+            {self.text_distance_metric.distance(field, "$1")} AS score
+          FROM embedding
+          WHERE collection_id = {self.collection_id}
+          ORDER BY {self.text_distance_metric.order_by(field, "$1")}
+        )
+        SELECT
+          relevant_embeddings.chunk_id AS chunk_id,
+          chunk.text AS text,
+          relevant_embeddings.score AS score
+        FROM relevant_embeddings
+        JOIN chunk
+        ON chunk.id = relevant_embeddings.chunk_id
+        LIMIT $2
```

Comment:
It seems like we could invert this and use `SELECT DISTINCT ... from chunk` to get the deduplicated chunks.

Reply:
Maybe? I want to get something in first and then play with it. I'd like to be able to point a pgsql repl at the database with chunks loaded in, and then see what works (and also use explain to see what the query does, etc.). Deferring.